src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -0,0 +1,192 @@
+    enqueue_locked(m);
+  }
+}
+
+// LogMessageBuffer consists of a multiple-part/multiple-line messsages.
+// the mutex here gurantees its integrity.
+void AsyncLogWriter::enqueue(LogFileOutput& output, LogMessageBuffer::Iterator msg_iterator) {
+  AsyncLogLocker lock;
+
+  for (; !msg_iterator.is_at_end(); msg_iterator++) {
+    AsyncLogMessage m(output, msg_iterator.decorations(), os::strdup(msg_iterator.message()));
+    enqueue_locked(m);
+  }
+}
+
+AsyncLogWriter::AsyncLogWriter()
+  : _state(ThreadState::NotReady),
+    _stats(17 /*table_size*/) {
+  if (os::create_thread(this, os::asynclog_thread)) {
+    _state = ThreadState::Initialized;
+  }
+
+  log_info(logging)("The maximum entries of AsyncLogBuffer: " SIZE_FORMAT ", estimated memory use: " SIZE_FORMAT " bytes",
+                    _buffer_max_size, AsyncLogBufferSize);
+}
+
+class AsyncLogMapIterator {
+  AsyncLogBuffer& _logs;
+
+ public:
+  AsyncLogMapIterator(AsyncLogBuffer& logs) :_logs(logs) {}
+  bool do_entry(LogFileOutput* output, uint32_t* counter) {
+    using none = LogTagSetMapping<LogTag::__NO_TAG>;
+
+    if (*counter > 0) {
+      LogDecorations decorations(LogLevel::Warning, none::tagset(), output->decorators());
+      stringStream ss;
+      ss.print(UINT32_FORMAT_W(6) " messages dropped due to async logging", *counter);
+      AsyncLogMessage msg(*output, decorations, ss.as_string(true /*c_heap*/));
+      _logs.push_back(msg);
+      *counter = 0;
+    }
+
+    return true;
+  }
+};
+
+void AsyncLogWriter::perform_IO() {
+  // Use kind of copy-and-swap idiom here.
+  // Empty 'logs' swaps the content with _buffer.
+  // Along with logs destruction, all procceeded messages are deleted.
+  //
+  // The operation 'pop_all()' is done in O(1). All I/O jobs are then performed without
+  // lock protection. This guarantees I/O jobs don't block logsites.
+  AsyncLogBuffer logs;
+  { // critical region
+    AsyncLogLocker ml;
+    AsyncLogMapIterator dropped_counters_iter(logs);
+
+    _buffer.pop_all(&logs);
+    // append meta-messages of dropped counters
+    _stats.iterate(&dropped_counters_iter);
+  }
+
+  LinkedListIterator<AsyncLogMessage> it(logs.head());
+  while (!it.is_empty()) {
+    AsyncLogMessage* e = it.next();
+    char* msg = e->message();
+
+    if (msg != nullptr) {
+      e->output()->write_blocking(e->decorations(), msg);
+      os::free(msg);
+    }
+  }
+}
+
+void AsyncLogWriter::run() {
+  assert(_state == ThreadState::Running, "sanity check");
+
+  while (_state == ThreadState::Running) {
+    _sem.wait();
+    perform_IO();
+  }
+
+  assert(_state == ThreadState::Terminated, "sanity check");
+  perform_IO(); // in case there are some messages left
+}
+
+AsyncLogWriter* AsyncLogWriter::_instance = nullptr;
+
+void AsyncLogWriter::initialize() {
+  if (!LogConfiguration::is_async_mode()) return;
+
+  if (_instance == nullptr) {
+    AsyncLogWriter* self = new AsyncLogWriter();
+
+    if (self->_state == ThreadState::Initialized) {
```

Comment:
but it's not exceptional safe. I know hotspot doesn't use c++ exception, but the general principle still applys. 
the more logics in ctor, the more likely to fail.  So far, if something wrong, I left self in state of `NotReady`.

Reply:
Since hotspot does not use c++ exception, I don't really follow the reasoning of "the more logics in ctor, the more likely to fail". Code doesn't just fail mysteriously. The only thing that could "fail" here is `new AsyncLogWriter()` returns `nullptr`, I think.

Anyway, this is just a suggestion; feel free to go for the style you prefer.